### ISSUE TITLE
AKS Connectivity issues with calico enabled and with no policies defined.

### ISSUE THREAD

**Issue by pranaypathik** [projectcalico/calico#4039]
## Current Behavior
Running Kubernetes on Azure via calico network cni enabled. 
With no calico policies defined we have issue whereby the pods loose connectivity to various resources.
This is resolved by restarting the calico typha pods.

Example of loss of connectivity:
- Azure Paas Example Azure keyvault or sql services. 
- Nginx ingress not able to resolve service end point giving 504 errors. 

Below is an example of error on a bad pod:
wget https://kv-redacted.vault.azure.net
wget: bad address 'kv-redacted.vault.azure.net'

on a good pod:
w get https://kv-redacted.vault.azure.net
Connecting to kv-redacted.vault.azure.net (13.72.250.239:443)
wget: server returned error: HTTP/1.1 403 Forbidden

Restarting calico typha deployment fixes the issue straight away.

Additional notes:
On calico typha logs this is reported:
2020-09-22 16:16:44.375 [INFO][7] watchercache.go 204: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/workloadendpoints" error=Get https://kv8s-cluster-url-redacted:443/api/v1/pods: dial tcp azure.ip.redacted:443: connect: connection refused

Checked that the ipaddress of the kubernetes master has not change.
Able to connect to the endpoint from a pod. Microsoft confirmed no issue at their end.
Sometimes the issue happens even with no timeout exceptions.

## Steps to Reproduce (for bugs)
The issue happens intermittently and does occurs mulitple kuberenetes instances.
I am not not able reproduce on demand.
Issue was raised with Microsoft. They have advised to raise a case with calico as they find no issue at their end.

## Context
Pods connectivity is impacted which causes various application issue. 
Have to manually intervene to resolve the issue.

## Your Environment
Orchestrator version: Kubernetes version: v1.17.7
Cloud Platform: Azure
Calico Version: v3.8.0

**Comment by pranaypathik** [comment]
Log files attached with sensitive data redacted
[2020.0923.logs.zip](https://github.com/projectcalico/calico/files/5302292/2020.0923.logs.zip)

**Comment by caseydavenport** [comment]
@pranaypathik looking at those logs for Typha, I see a few things.

First, some logs like this:

```
1902 E0922 16:16:42.259328       7 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=144699, ErrCode=NO_ERROR, debug=""
```

Then, some logs like this:

```
1909 2020-09-22 16:16:43.232 [INFO][7] watchercache.go 259: Failed to create watcher ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=Get https://kv8s-cluster-url-redacted:443/api/v1/nodes?resourceVersion=44610930&watch=true: dial tcp azure.ip.redacted:443: connect: connection refused performFullResync=false
```

These both suggest an issue with Typha being unable to connect to the Kubernetes API server. 

Given Typha restarts fix the issue, it sounds like a problem with the connection on the Typha side rather than the API server side. 

It looks like you might be hitting this issue: https://github.com/projectcalico/libcalico-go/issues/1267

The root cause is that the Kubernetes client and golang HTTP/2 libraries don't detect when the connection is unexpected severed, and so Calico isn't informed that the connection has gone down.

There is an upstream fix that we're waiting on at the moment here: https://github.com/kubernetes/kubernetes/pull/95981

**Comment by caseydavenport** [comment]
For now, I'm going to close this as a duplicate of https://github.com/projectcalico/libcalico-go/issues/1267 since I believe they are the same root cause.
